### ISSUE TITLE
pyenv: vowpalwabbit python wrappers do not build bc "Could NOT find PythonLibs" on Ubuntu 18.04 LTS

### ISSUE THREAD

**Issue by abarbosa94** [VowpalWabbit/vowpal_wabbit#2504]
#### Describe the bug
The steps described in https://vowpalwabbit.org/start.html do not work in a `pyenv` environment on Ubuntu 18.04.4 LTS

#### To Reproduce
Just copied the commands provided from the website:
```
sudo apt update

sudo apt install libboost-dev libboost-program-options-dev libboost-system-dev libboost-thread-dev libboost-math-dev libboost-test-dev libboost-python-dev zlib1g-dev cmake python3 python3-pip

pip3 install vowpalwabbit

```

#### Expected behavior
vowpalwabbit did not get installed ):

#### Observed Behavior
```
Collecting vowpalwabbit                                                                                                                                                                                     
  Downloading vowpalwabbit-8.8.1.tar.gz (3.1 MB)                                                                                                                                                            
     |████████████████████████████████| 3.1 MB 2.2 MB/s                                                                                                                                                     
Building wheels for collected packages: vowpalwabbit                                                                                                                                                        
  Building wheel for vowpalwabbit (setup.py) ... error                                                                                                                                                      
  ERROR: Command errored out with exit status 1:                                                                                                                                                            
   command: /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/bin/python3.7 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-2u0tfiy7/vowpalwabbit/setup.py'"'"'; __file__$
'"'"'/tmp/pip-install-2u0tfiy7/vowpalwabbit/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"$
"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-gmx6xhef                                                                                                                                                       
       cwd: /tmp/pip-install-2u0tfiy7/vowpalwabbit/                                                                                                                                                         
  Complete output (61 lines):                                                                                                                                                                               
  running bdist_wheel                                                                                                                                                                                       
  running build                                                                                                                                                                                             
  running build_py                                                                                                                                                                                          
  creating build                                                                                                                                                                                            
  creating build/lib.linux-x86_64-3.7                                                                                                                                                                       
  creating build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                                                          
  copying python/vowpalwabbit/sklearn_vw.py -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                      
  copying python/vowpalwabbit/__init__.py -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                        
  copying python/vowpalwabbit/pyvw.py -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                            
  running egg_info                                                                                                                                                                                          
  writing python/vowpalwabbit.egg-info/PKG-INFO                                                                                                                                                             
  writing dependency_links to python/vowpalwabbit.egg-info/dependency_links.txt                                                                                                                             
  writing top-level names to python/vowpalwabbit.egg-info/top_level.txt                                                                                                                                     
  reading manifest file 'python/vowpalwabbit.egg-info/SOURCES.txt'                                                                                                                                          
  reading manifest template 'MANIFEST.in'                                                                                                                                                                   
  warning: no files found matching 'python/*.dll' under directory 'python/bin'                                                                                                                              
  warning: no files found matching 'python/packages.config' under directory 'python/windows*'                                                                                                               
  warning: no files found matching 'python/*.vcxproj*' under directory 'python/windows*'                                                                                                                    
  warning: no previously-included files matching '*.o' found anywhere in distribution                                                                                                                       
  warning: no previously-included files matching '*.exe' found anywhere in distribution                                                                                                                     
  warning: no previously-included files matching '*.pyc' found anywhere in distribution                                                                                                                     
  warning: no previously-included files matching '*' found under directory 'vowpalwabbit/.nuget'                                                                                                            
  warning: no previously-included files matching '*' found under directory 'vowpalwabbit/slim'                                                                                                              
  writing manifest file 'python/vowpalwabbit.egg-info/SOURCES.txt'                                                                                                                                          
  copying python/vowpalwabbit/conda_install.sh -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                   
  running build_ext                                                                                                                                                                                         
  creating build/temp.linux-x86_64-3.7                                                                                                                                                                      
  cmake -DCMAKE_BUILD_TYPE=Release -DPY_VERSION=3.7 -DBUILD_PYTHON=On -DBUILD_TESTS=Off -DWARNINGS=Off -DBoost_NO_BOOST_CMAKE=ON -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/tmp/pip-install-2u0tfiy7/vowpalwabbit/bui
ld/lib.linux-x86_64-3.7 /tmp/pip-install-2u0tfiy7/vowpalwabbit                                                                                                                                              
  -- VowpalWabbit Version: 8.8.1                                                                                                                                                                            
  -- The CXX compiler identification is GNU 7.5.0                                                                                                                                                           
  -- Check for working CXX compiler: /usr/bin/c++                                                                                                                                                           
  -- Check for working CXX compiler: /usr/bin/c++ - works                                                                                                                                                   
  -- Detecting CXX compiler ABI info                                               
  -- Detecting CXX compile features - done                                                                                                                                                                  
  -- Number of processors: 8                                                                                                                                                                                
  -- Looking for C++ include pthread.h                                                                                                                                                                      
  -- Looking for C++ include pthread.h - found                                                                                                                                                              
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD                                                                                                                                                                
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed                                                                                                                                                       
  -- Check if compiler accepts -pthread                                                                                                                                                                     
  -- Check if compiler accepts -pthread - yes                                                                                                                                                               
  -- Found Threads: TRUE                                                                                                                                                                                    
  -- Found Boost: /usr/include (found version "1.65.1") found components: program_options system thread unit_test_framework chrono date_time atomic                                                         
  -- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")                                                                                                                                 
  -- Python 3.7 will be used for VowpalWabbit Python bindings                                                                                                                                               
  -- Found PythonInterp: /home/andrebarbosa/.pyenv/versions/jon-snow/bin/python3.7 (found suitable version "3.7.2", minimum required is "3.7")                                                              
  CMake Error at /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):                  
    Could NOT find PythonLibs (missing: PYTHON_LIBRARIES PYTHON_INCLUDE_DIRS)                                                                                                                               
    (Required is at least version "3.7")                                                                                                                                                                    
  Call Stack (most recent call first):                                                                                                                                                                      
    /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:445 (_FPHSA_FAILURE_MESSAGE)                 
    /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.17/Modules/FindPythonLibs.cmake:310 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)                     
    python/CMakeLists.txt:22 (find_package)                                                                                                                                                                 
                                                                                                                                                                                                            
                                                                                                                                                                                                            
  -- Configuring incomplete, errors occurred!                                                                                                                                                               
  See also "/tmp/pip-install-2u0tfiy7/vowpalwabbit/build/temp.linux-x86_64-3.7/CMakeFiles/CMakeOutput.log".                                                                                                 
  See also "/tmp/pip-install-2u0tfiy7/vowpalwabbit/build/temp.linux-x86_64-3.7/CMakeFiles/CMakeError.log".                                                                                                  
  error: command 'cmake' failed with exit status 1                                                                                                                                                          
  ----------------------------------------                                                                                                                                                                  
  ERROR: Failed building wheel for vowpalwabbit  
```

#### Environment
VW Version 8.8.1 on Ubuntu 18.04.4 LTS inside a `pyenv` environment with the Python bindings


#### Additional context
I have tried to install it from source to see if works and I got the following error:
```
/usr/bin/ld: /home/andrebarbosa/.pyenv/versions/3.7.2/lib/libpython3.7m.a(ceval.o): relocation R_X86_64_PC32 against symbol `_PyRuntime' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: Bad value
collect2: error: ld returned 1 exit status
python/CMakeFiles/pylibvw.dir/build.make:103: recipe for target '../lib.linux-x86_64-3.7/pylibvw.so' failed
make[3]: *** [../lib.linux-x86_64-3.7/pylibvw.so] Error 1
CMakeFiles/Makefile2:798: recipe for target 'python/CMakeFiles/pylibvw.dir/all' failed
make[2]: *** [python/CMakeFiles/pylibvw.dir/all] Error 2
CMakeFiles/Makefile2:810: recipe for target 'python/CMakeFiles/pylibvw.dir/rule' failed
make[1]: *** [python/CMakeFiles/pylibvw.dir/rule] Error 2
Makefile:372: recipe for target 'pylibvw' failed
make: *** [pylibvw] Error 2
error: command 'cmake' failed with exit status 2

```

**Comment by jackgerrits** [comment]
Looks like you're missing the Python dev dependencies - 

Can you try running:
```
sudo apt install python3-dev
```

**Comment by lalo** [comment]
If that doesn't work, try also specifying the environment variable CMAKE_PREFIX_PATH to the corresponding pyenv `.pyenv/versions/<venv>/` directory and then retry.

You want to make sure that cmake finds the corresponding PythonLibs
```
CMake Error at /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):                  
    Could NOT find PythonLibs (missing: PYTHON_LIBRARIES PYTHON_INCLUDE_DIRS)                                                                                                                               
    (Required is at least version "3.7")        
```

**Comment by abarbosa94** [comment]
Thanks for quick responses!



> Looks like you're missing the Python dev dependencies -
> 
> Can you try running:
> 
> ```
> sudo apt install python3-dev
> ```

I have done it :) Errors are still the same



> If that doesn't work, try also specifying the environment variable CMAKE_PREFIX_PATH to the corresponding pyenv `.pyenv/versions/<venv>/` directory and then retry.
> 
> You want to make sure that cmake finds the corresponding PythonLibs
> 
> ```
> CMake Error at /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.17/Modules/FindPackageHandleStandardArgs.cmake:164 (message):                  
>     Could NOT find PythonLibs (missing: PYTHON_LIBRARIES PYTHON_INCLUDE_DIRS)                                                                                                                               
>     (Required is at least version "3.7")        
> ```

I am not so familiar with C++ code. How am I supposed to do this? Just by something like:
`export CMAKE_PREFIX_PATH=/home/andrebarbosa/.pyenv/versions/<venv>/` ?

**Comment by lalo** [comment]
I'm not that familiar with how pyenv structures the dirs but one of these might work? Try in the following order

`export CMAKE_PREFIX_PATH=/home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/`
or
`export CMAKE_PREFIX_PATH=/home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/`
or
`export CMAKE_PREFIX_PATH=/home/andrebarbosa/.pyenv/versions/3.7.2/envs/`
or
`export CMAKE_PREFIX_PATH=/home/andrebarbosa/.pyenv/versions/3.7.2/`

after doing that on the same terminal run again the command that triggers cmake which I assume is the pip install.

I would also delete this env var after CMAKE_PREFIX_PATH, to avoid it interrupting any other environment that uses cmake.

do this by running

`unset CMAKE_PREFIX_PATH`

**Comment by abarbosa94** [comment]
Thanks for the quick answers! However, I still got the same problem ):

Any other hypothesis?

**Comment by lalo** [comment]
Was it exactly the same error though?

Anyways, try running 

`which python` copy the result and report it back here.
`which python3` copy the result and report it back here.
`which pip3` copy the result and report it back here.

Then run
`pyenv shell system` and this should run successfully, report back if it fails with the error.

Then run again:
`which python` copy the result and report it back here.
`which python3` copy the result and report it back here.
`which pip3` copy the result and report it back here.

They _should_ be a different from the first set, but report them anyways to see how they differ.

Try running then `pip3 install vowpalwabbit` and report back again whatever error if any.

The thing that is missing from your repro is that you mentioned you ran the 3 commands, but in no where is pyenv being mentioned on how it was setup/activated.

**Comment by abarbosa94** [comment]
Sorry about late answer
> which python copy the result and report it back here.

`/home/andrebarbosa/.pyenv/shims/python`

> which python3 copy the result and report it back here.

`/home/andrebarbosa/.pyenv/shims/python3`

> which pip3 copy the result and report it back here.

`/home/andrebarbosa/.pyenv/shims/pip3`

> pyenv shell system and this should run successfully, report back if it fails with the error.

Indeed, no errors :)

> which python copy the result and report it back here.

`/home/andrebarbosa/.pyenv/shims/python`

> which python3 copy the result and report it back here.

`/home/andrebarbosa/.pyenv/shims/python3`

> which pip3 copy the result and report it back here.

`/home/andrebarbosa/.pyenv/shims/pip3`

This was unexpected though. The output was exactly the same ):


> Try running then pip3 install vowpalwabbit and report back again whatever error if any.

By following this, it works! However, as expected, it has installed the vowpalwabbit under the `system` environment, while I am trying to installing it under a specific virtualenv (`jon-snow`)

Then, if I run the following:
```
pyenv shell jon-snow
pip3 install vowpalwabbit
```

I got the following error, which is the same that I was getting before:
```
  Building wheel for vowpalwabbit (setup.py) ... error                                                                                                                                                      
  ERROR: Command errored out with exit status 1:                                                                                                                                                            
   command: /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/bin/python3.7 -u -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-install-akttgglt/vowpalwabbit/setup.py'"'"'; __file__=
'"'"'/tmp/pip-install-akttgglt/vowpalwabbit/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'
"'exec'"'"'))' bdist_wheel -d /tmp/pip-wheel-_n1qwns6                                                                                                                                                       
       cwd: /tmp/pip-install-akttgglt/vowpalwabbit/                                                                                                                                                         
  Complete output (60 lines):                                                                                                                                                                               
  running bdist_wheel                                                                                                                                                                                       
  running build                                                                                                                                                                                             
  running build_py                                                                                                                                                                                          
  creating build                                                                                                                                                                                            
  creating build/lib.linux-x86_64-3.7                                                                                                                                                                       
  creating build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                                                          
  copying python/vowpalwabbit/sklearn_vw.py -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                      
  copying python/vowpalwabbit/__init__.py -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                        
  copying python/vowpalwabbit/pyvw.py -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                            
  running egg_info                                                                                                                                                                                          
  writing python/vowpalwabbit.egg-info/PKG-INFO                                                                                                                                                             
  writing dependency_links to python/vowpalwabbit.egg-info/dependency_links.txt                                                                                                                             
  writing top-level names to python/vowpalwabbit.egg-info/top_level.txt                                                                                                                                     
  reading manifest file 'python/vowpalwabbit.egg-info/SOURCES.txt'                                                                                                                                          
  reading manifest template 'MANIFEST.in'                                                                                                                                                                   
  warning: no files found matching 'python/*.dll' under directory 'python/bin'                                                                                                                              
  warning: no files found matching 'python/packages.config' under directory 'python/windows*'                                                                                                               
  warning: no files found matching 'python/*.vcxproj*' under directory 'python/windows*'                                                                                                                    
  warning: no previously-included files matching '*.o' found anywhere in distribution                                                                                                                       
  warning: no previously-included files matching '*.exe' found anywhere in distribution                                                                                                                     
  warning: no previously-included files matching '*.pyc' found anywhere in distribution                                                                                                                     
  warning: no previously-included files matching '*' found under directory 'vowpalwabbit/.nuget'                                                                                                            
  warning: no previously-included files matching '*' found under directory 'vowpalwabbit/slim'                                                                                                              
  writing manifest file 'python/vowpalwabbit.egg-info/SOURCES.txt'                                                                                                                                          
  copying python/vowpalwabbit/conda_install.sh -> build/lib.linux-x86_64-3.7/vowpalwabbit                                                                                                                   
  running build_ext                                                                                                                                                                                         
  creating build/temp.linux-x86_64-3.7
  cmake -DCMAKE_BUILD_TYPE=Release -DPY_VERSION=3.7 -DBUILD_PYTHON=On -DBUILD_TESTS=Off -DWARNINGS=Off -DBoost_NO_BOOST_CMAKE=ON -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=/tmp/pip-install-akttgglt/vowpalwabbit/bu$
ld/lib.linux-x86_64-3.7 /tmp/pip-install-akttgglt/vowpalwabbit
  -- VowpalWabbit Version: 8.8.1
  -- The CXX compiler identification is GNU 7.5.0
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Number of processors: 8
  -- Looking for C++ include pthread.h
  -- Looking for C++ include pthread.h - found
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD
  -- Performing Test CMAKE_HAVE_LIBC_PTHREAD - Failed
  -- Check if compiler accepts -pthread
 -- Check if compiler accepts -pthread - yes
  -- Found Threads: TRUE
  -- Found Boost: /usr/include (found version "1.65.1") found components: program_options system thread unit_test_framework chrono date_time atomic
  -- Found ZLIB: /usr/lib/x86_64-linux-gnu/libz.so (found version "1.2.11")
  -- Python 3.7 will be used for VowpalWabbit Python bindings
  -- Found PythonInterp: /home/andrebarbosa/.pyenv/versions/jon-snow/bin/python3.7 (found suitable version "3.7.2", minimum required is "3.7")
  CMake Error at /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:165 (message):
    Could NOT find PythonLibs (missing: PYTHON_LIBRARIES PYTHON_INCLUDE_DIRS)
    (Required is at least version "3.7")
  Call Stack (most recent call first):
    /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.18/Modules/FindPackageHandleStandardArgs.cmake:458 (_FPHSA_FAILURE_MESSAGE)
    /home/andrebarbosa/.pyenv/versions/3.7.2/envs/jon-snow/lib/python3.7/site-packages/cmake/data/share/cmake-3.18/Modules/FindPythonLibs.cmake:310 (FIND_PACKAGE_HANDLE_STANDARD_ARGS)
    python/CMakeLists.txt:22 (find_package)
  
  
  -- Configuring incomplete, errors occurred!
  See also "/tmp/pip-install-akttgglt/vowpalwabbit/build/temp.linux-x86_64-3.7/CMakeFiles/CMakeOutput.log".
  See also "/tmp/pip-install-akttgglt/vowpalwabbit/build/temp.linux-x86_64-3.7/CMakeFiles/CMakeError.log".
  error: command 'cmake' failed with exit status 1
```

>The thing that is missing from your repro is that you mentioned you ran the 3 commands, but in no where is pyenv being mentioned on how it was setup/activated.

I ran the following set of commands:

```
        pyenv install -s 3.7.2                                                                                                                                                                              
        pyenv virtualenv 3.7.2 jon-snow                                                                                                                                                                     
        pyenv local jon-snow  
```

**Comment by lalo** [comment]
We could probably beef up our cmake scripts to be able to find the python libraries, pyenv seems to behave differently or the PythonLibs finder script hasn't been updated (if this is the case). At least is good to hear that you were able to install it on the local python, this is good info for further root-causing the underlying issue.

**Comment by jackgerrits** [comment]
There are now binary wheels available for Linux, including the 8.8.1 release. If you try this again, it should now use the binary wheel. If this still doesn't work please reopen the issue. Thanks!
